[FIX] website: fix the recaptcha legal notice option

tdecaluwe opened the pull request:
When adding an editable node to a non-editable region (which in turn is situated in an editable part of the document), the editor triggers a rollback because the new node is considered an editable root, making it unremovable.

However, semantically, the new node is not an editable root, being a descendant of another node having the `contenteditable` property set to true. Logically, it also makes sense to consider these elements as removable.

The rollback is currently triggered when adding an editable node inside a non editable part of a snippet, for example when toggling the recaptcha legal notice on a form.

A commit was cherry picked from pull request #86489, simply using the special root `oid` as a check for editable roots.

opw-2723017

tdecaluwe commented:
@dmo-odoo Another bug caused by an unwanted rollback in the OdooEditor (similar to https://github.com/odoo/odoo/pull/87519 but different). Bommits resolve the problem. The first one is a hacky workaround (see cef4f6080c1dfce9295b9e6764b959ab7c6020e6). However, solving the root cause (see db8fba504948b4d65de6252289ba1c848ee194bb) results in failing editor tests. Could you verify if the logic of the fix makes sense at least?

Zinston commented:
> However, solving the root cause (see https://github.com/odoo/odoo/commit/db8fba504948b4d65de6252289ba1c848ee194bb) results in failing editor tests

@tdecaluwe Looking at the errors I think it should be fixed by using `closestElement(node, 'selector')` rather than `node.closest('selector')` (you might need to import `closestElement` from utils) because `closest` doesn't exist on text nodes.

dmo-odoo commented:
I think it is our unremovable rollback detection logic that should be fixed so I would prefer the proper solution at the root cause if possible. However, I had a quick look at the test cases that the change broke and I don't think the tests are wrong. Any chance `closestElement` finds a editable container too high in the test environment and gets confused ?

tdecaluwe commented:
@dmo-odoo You're right, the tests weren't the problem. Actually `closest` also considers the element itself, not only the ancestors, meaning it also returned an element for an actual editable root. I fixed the problem and also added a test using `historyStep`, just like the problem in the bug report.

tdecaluwe commented on addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:
```diff
@@ -965,7 +965,11 @@ export function isUnremovable(node) {
+        // of another editable element. Only an empty attribute or an attribute
+        // set to true will render an element editable. Nodes having an invalid
+        // contenteditable value will inherit the property from their parents.
+        !closestElement(node.parentElement, '[contenteditable=""], [contenteditable="true"]') &&
```
Looks like it would work, I'll test and confirm.

robodoo commented:
@dmo-odoo, you may want to rebuild or fix this PR as it has failed CI.

robodoo commented:
Merge method set to rebase and fast-forward

odony commented:
@robodoo override=ci/security
R: test code using hardcoded innerHTML

Xavier-Do commented:
We could grep for `\/test(s?)\/` in filepath and skip them
Is that reliable? Enough for this ci I guess. But you can still add a static/tests/file.js in base assets
